### PR TITLE
Issue 1012: Implement welcome page for logged-in users

### LINKED ISSUE #1012: Implement Welcome Page for Logged In Users
Implement a basic welcome page for logged in users.

**Acceptance Criteria:**
- [ ] Matches UX designs (link)
- [ ] Responds correctly to user role / lack thereof

### PR DESCRIPTION
**Pull request by jorgegonzalez**
### Summary of Changes

This PR adds a proper Welcome page and frontend route at `/welcome` to the app.

_Addresses issue #1012
Closes #1012
_Acceptance criteria as stated in the issue_

### How to Test
_List the steps to test the PR_

- Check that after logging in, the user is redirected to the new Welcome page.
- Check that the welcome page displays the current user's role and permissions (if any), or a message saying the user's "request for access is currently being reviewed".
	- To test both cases, use Django Admin to add/remove roles

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
	- [ ] Matches UX designs (link)
	- [ ] Responds correctly to user role / lack thereof

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: 90% (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces frontend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
	- No issues found

### CODE REVIEW

File: tdrs-frontend/src/components/LoginCallback/LoginCallback.jsx [hunk 1]
```diff
@@ -40,7 +40,7 @@ function LoginCallback() {
     return <Redirect to="/" />
   }
 
-  return authenticated ? <Redirect to="/edit-profile" /> : null
+  return authenticated ? <Redirect to="/welcome" /> : null
```
Comment (jorgegonzalez) [inline tdrs-frontend/src/components/LoginCallback/LoginCallback.jsx:43]: Updated the login callback to redirect users to the new Welcome page.

File: tdrs-backend/wait_for_postgres.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 
 def pg_isready(host, user, password, dbname):
     """Check to see if Postgres is ready."""
-    while time() - start_time < check_timeout:
+    while time() - start_time < float(check_timeout):
```
Comment (jorgegonzalez) [inline tdrs-backend/wait_for_postgres.py:27]: This is an unrelated fix for a local development error that occurs when using the `POSTGRES_CHECK_TIMEOUT` env var.

File: tdrs-frontend/src/components/Routes/Routes.js [hunk 3]
```diff
@@ -22,15 +24,18 @@ const Routes = () => {
       <Route exact path="/login">
         <LoginCallback />
       </Route>
+      <PrivateRoute exact title="Welcome to TDP" path="/welcome">
```
Comment (jorgegonzalez) [inline tdrs-frontend/src/components/Routes/Routes.js:27]: Added discrete `/welcome` route.

File: tdrs-frontend/src/components/Welcome/Welcome.jsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import React from 'react'
+import Button from '../Button'
+import { faSignOutAlt } from '@fortawesome/free-solid-svg-icons'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { useSelector } from 'react-redux'
+
+function Welcome() {
+  const logOut = () => {
+    window.location = `${process.env.REACT_APP_BACKEND_URL}/logout/oidc`
```
Comment (jorgegonzalez) [inline tdrs-frontend/src/components/Welcome/Welcome.jsx:9]: This is the identical behavior of our other logout functions in the code base.

File: tdrs-frontend/src/components/Welcome/Welcome.jsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          <p className="margin-top-1 margin-bottom-4">
+            You've been approved as a/an {role?.[0].name}. You'll be able to do
+            the following in TDP:
+          </p>
+          <ul>
+            {role?.[0]?.permissions?.map((permission) => (
+              <li key={permission.id} id={permission.id}>
+                {permission.name}
+              </li>
+            ))}
```
Comment (jorgegonzalez) [inline tdrs-frontend/src/components/Welcome/Welcome.jsx:33]: The design for this will likely change but this currently shows a list of the user's permissions on the Welcome page.

File: tdrs-backend/wait_for_postgres.py (hunk 2 above)
Comment (abottoms-coder) [inline tdrs-backend/wait_for_postgres.py:27]: Could you please comment on the context for this change?

**Review by abottoms-coder — APPROVED** [review]

File: .circleci/config.yml [hunk 6]
```diff
@@ -401,6 +401,9 @@ jobs:
       - store_artifacts:
           path: tdrs-frontend/pa11y-screenshots/data-files.png
 
+      - store_artifacts:
+          path: tdrs-frontend/pa11y-screenshots/welcome.png
```
Comment (jtwillis92) [inline .circleci/config.yml:405]: I think we could pass the whole directory to `path` here and not have to define every PNG.

File: tdrs-frontend/src/components/Welcome/Welcome.jsx (hunk 4 above)
Comment (jtwillis92) [reply tdrs-frontend/src/components/Welcome/Welcome.jsx:9]: Minor nit but if we have multiple functions doing this same thing it would be a good candidate for a utils function that could be imported.

File: tdrs-frontend/src/components/Welcome/Welcome.jsx (hunk 5 above)
Comment (jtwillis92) [reply tdrs-frontend/src/components/Welcome/Welcome.jsx:33]: This might not show as much data until #1230 is merged but I don't think that is an issue at this time.

File: tdrs-frontend/src/components/Welcome/Welcome.jsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    <div className="margin-top-5">
+      {!hasRole ? (
+        <p className="margin-top-1 margin-bottom-4">
+          Your request for access is currently being reviewed by an OFA Admin.
+          We'll send you an email when it's been approved.
+        </p>
+      ) : (
+        <>
+          <p className="margin-top-1 margin-bottom-4">
+            You've been approved as a/an {role?.[0].name}. You'll be able to do
```
Comment (jtwillis92) [inline tdrs-frontend/src/components/Welcome/Welcome.jsx:25]:
```suggestion
            You've been approved as a(n) {role?.[0].name}. You'll be able to do
```

**Review by jtwillis92 — APPROVED** [review]
Left a couple nits/suggestions, but I was able to run through the testing steps and confirmed this works as expected.

**Comment by jorgegonzalez** [comment]
Here are screenshots as a logged-in user, without and with a role (OFA Admin).

image
Screen Shot 2021-08-24 at 9 36 54 AM

**Comment by ADPennington** [comment]
@jorgegonzalez test results 👍🏾 --all the redirects to welcome page show the expected permissions 👍🏾 but I'm perplexed by the sys admin groups results, so asyncing w/ @jtwillis92 because it's beyond the scope of this PR. 

- `data analyst` login
1012p1

- `ofa admin` login ✔️ 
1012p4

- `ofa sys admin` login ❔ 
1012p2

- `sys admin` login ❔ 
1012p3

- login w/ no role assigned ✔️ 
1012p5

- login with `deactivated` flag checked ✔️ 
1012p6

**Review by ADPennington — APPROVED** [review]
thanks @jorgegonzalez 👍🏾 @abottoms-coder ready to merge

---
### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] Matches UX designs (link)
  + [x] Responds correctly to user role / lack thereof

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?

### Deliverable 4: Accessible

_no a11y review yet, as the welcome page is not yet in its final form. content to be added and formatted in the future PR_


### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no new issues detected. good to see welcome page png updated!_
